server/src/main/java/org/jboss/as/server/controller/git/ElytronClientSshdSessionFactory.java

```diff
@@ -0,0 +1,226 @@
+import org.eclipse.jgit.transport.sshd.SshdSessionFactory;
+import org.eclipse.jgit.util.FS;
+import org.jboss.as.server.logging.ServerLogger;
+import org.wildfly.client.config.ConfigXMLParseException;
+import org.wildfly.security.auth.callback.CredentialCallback;
+import org.wildfly.security.auth.client.AuthenticationConfiguration;
+import org.wildfly.security.auth.client.AuthenticationContext;
+import org.wildfly.security.auth.client.AuthenticationContextConfigurationClient;
+import org.wildfly.security.auth.client.ElytronXmlParser;
+import org.wildfly.security.credential.KeyPairCredential;
+import org.wildfly.security.credential.PasswordCredential;
+import org.wildfly.security.credential.SSHCredential;
+import org.wildfly.security.password.interfaces.ClearPassword;
+
+
+/**
+ * SshdSessionFactory for JGIt using Elytron.
+ * Supports specifying ssh-keys as {@link KeyPairCredential} or {@link SSHCredential}
+ *
+ * @author <a href="mailto:aabdelsa@redhat.com">Ashley Abdel-Sayed</a>
+ */
+public class ElytronClientSshdSessionFactory extends SshdSessionFactory {
+
+    private static AuthenticationContextConfigurationClient CLIENT = AccessController.doPrivileged(AuthenticationContextConfigurationClient.ACTION);
+    private final AuthenticationContext context;
+    private final Map<Tuple, Iterable<KeyPair>> defaultKeys;
+    private String[] defaultIdentities;
+    private KeyPair keyPair;
+    private URI uri;
+    private String knownHostsFile;
+
+    public ElytronClientSshdSessionFactory(URI authenticationConfig) throws ConfigXMLParseException, GeneralSecurityException {
+        context = ElytronXmlParser.parseAuthenticationClientConfiguration(authenticationConfig).create();
+        this.defaultKeys = new ConcurrentHashMap();
+        this.defaultIdentities = SshConstants.DEFAULT_IDENTITIES;
+        this.keyPair = null;
+        this.knownHostsFile = "known_hosts";
+    }
+
+    @Override
+    @NonNull
+    protected List<Path> getDefaultIdentities(@NonNull File sshDir) {
+        return (List) Arrays.asList(this.defaultIdentities).stream().map((s) -> {
+            return (new File(sshDir, s)).toPath();
+        }).filter((var0) -> {
+            return Files.exists(var0, new LinkOption[0]);
+        }).collect(Collectors.toList());
+    }
+
+    @Override
+    public SshdSession getSession(URIish urIish, CredentialsProvider credentialsProvider, FS fs, int tms) throws TransportException {
+        try {
+            this.uri = urIishToUri(urIish);
+            AuthenticationConfiguration config = CLIENT.getAuthenticationConfiguration(this.uri, context);
+            CredentialCallback locationCallback = new CredentialCallback(SSHCredential.class, null);
+            CredentialCallback keyPairCallback = new CredentialCallback(KeyPairCredential.class, null);
+            Callback[] callbacks = {locationCallback, keyPairCallback};
+            CLIENT.getCallbackHandler(config).handle(callbacks);
```

Comment:
I think you could first try to handle the ```locationCallback``` and then if ```locationCallback.getCredential()``` returns null, then try to handle the other callback.

Reply:
I've kept that it always checks both credentials for the case where the user needs to use the KeyPairCredential to specify their keys and also use the SSHCredential to specify the location and filename of their known hosts file. Probably a rare case but I thought to keep it just in case